codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2RemoteFlowController.java

```diff
@@ -232,21 +232,27 @@ private void writePendingBytes() {
         int connectionWindow = state(connectionStream).window();
 
         if (connectionWindow > 0) {
-            writeChildren(connectionStream, connectionWindow);
-            for (Http2Stream stream : connection.activeStreams()) {
-                writeChildNode(state(stream));
+            // Allocate the bytes for the connection window to the streams, but do not write.
+            allocateBytesForTree(connectionStream, connectionWindow);
+
+            // Now write all of the allocated bytes. Copying the activeStreams array to avoid
+            // side effects due to stream removal/addition which might occur as a result
+            // of end-of-stream or errors.
+            for (Http2Stream stream : connection.activeStreams().toArray(new Http2Stream[0])) {
```

Comment:
Depending on the ratio of how many streams get bytes allocated vs how many active stream there are it may be beneficial to maintain a set member variable to track which streams have bytes allocated (or need to write) for this pass....I think I was previously passing a set as an argument though the recursive calls...and this may have been before I change the structure to what it is now....